src/core/net/ip6.cpp

```diff
@@ -977,6 +978,13 @@ Error Ip6::PassToHost(Message           &aMessage,
     // Do not pass IPv6 packets that exceed kMinimalMtu.
     VerifyOrExit(aMessage.GetLength() <= kMinimalMtu, error = kErrorDrop);
 
+    // When the packet is classified as 'receive' or it has a multicast destination we should always forward it to host.
+    // When the packet's source is mesh-local and the destination is not on the host, we should not forward it to host.
+    VerifyOrExit(aReceive || aMessageInfo.GetSockAddr().IsMulticast() ||
+                     !Get<Mle::Mle>().IsMeshLocalAddress(aMessageInfo.GetPeerAddr()) ||
+                     Get<ThreadNetif>().GetUnicastAddresses().ContainsMatching(aMessageInfo.GetSockAddr()),
```

Comment:
Can use `HasUnicastAddress()` instead of `GetUnicastAddresses().ContainsMatching()`